### ISSUE TITLE
[ONNX]  exported nodes of Multi-head attention can be simplified

### ISSUE THREAD

**Issue by m23ayou2** [pytorch/pytorch#151209]
I am exporting the nn.multiheadattention layer from pytorch to onnx and i have seen that many new operations that are not expected 

![Image](https://github.com/user-attachments/assets/7b37f417-23af-4def-8c7a-d5383e827fe0)

Is it a bug or a feature!

**Comment by justinchuby** [comment]
Could you explain your expectation?

**Comment by m23ayou2** [comment]
this is the wrapper around the nn.multiheadattention i used, 

```
class mha_(nn.Module): 
    def __init__(self): 
        super(mha_, self).__init__()

        self.li = nn.Linear(embed_dim, embed_dim)

        self.ma = nn.MultiheadAttention(embed_dim=embed_dim, num_heads=num_heads, dropout=0, bias=False, add_bias_kv=False, add_zero_attn=False)

    def forward(self, x):
        li = self.li(x)
        q = li
        k = li
        v = li

        output = self.ma(q, k, v)
        return output
mha_c = mha_()
```

i rebuild the model and was expecting something like this: 

![Image](https://github.com/user-attachments/assets/50027311-d22a-4b47-b229-87ef0d18a019)

**Comment by justinchuby** [comment]
Could you attach a script that will reproduce the onnx model (with how you call export)? If you have access to the expected model we can look deeper at that too.

**Comment by m23ayou2** [comment]
For sure, I joined a GitHub repo with a Jupyter notebook that reproduces the same exported model. The requirements file is there too. The visualization was done with Netron.

https://github.com/m23ayou2/Pytorch_issue_mha_onnx

Hope this helps.

**Comment by justinchuby** [comment]
Thanks. It appears to be due to pytorch's implementation of multi-head attention. In particular, its qkv matrix is packed into a single tensor (and your reference is not), which is why you would see a different computation graph (as well as gather nodes).

![Image](https://github.com/user-attachments/assets/98053c61-1a4b-4d95-baed-c2bda10f583f)

This part may be optimized away if we think about it carefully... Please feel free to suggest what you think can work as a graph rewrite rule.

**Comment by justinchuby** [comment]
Also:

Please use `torch.onnx.export(..., dynamo=True)` using the latest torch-nightly when exporting.

**Comment by m23ayou2** [comment]
If we overlook the other parts of the model and optimize only for the graph snippet, I think we necessarily need a reshape (go from a 3-dim tensor to a 4-dim tensor) and a permute to separate the different values for later for (q, k, v). We could probably do better if we looked into the gather and reshape operations that follow. This is an optimized version from 4 to 2 operations on the optimize jupyter notebook

I joined in the same GitHub repo and test it on a tensors: 

https://github.com/m23ayou2/Pytorch_issue_mha_onnx/blob/main/optimize.ipynb

**Comment by justinchuby** [comment]
cc @gramalingam @titaiwangms for optimization opportunities

**Comment by titaiwangms** [comment]
@m23ayou2 Could you try `torch.onnx.export(..., dynamo=True)` like @justinchuby suggested. The attachment is what I got with the new exporter. It looks pretty close to what you expected if I am not mistaken. Also, we have opset23 Attention support that help us get rid of all these operators (you might be interested.) in torch nightly (2.8 will soon be released).  Op spec: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Attention

[mha_models.zip](https://github.com/user-attachments/files/20843421/mha_models.zip)

**Comment by m23ayou2** [comment]
The computational graph you have joined  is substantially better than the first results and authentically reproduce what i was expecting, but could not reproduce it just using `torch.onnx.export(..., dynamo=True)
` If you can attach the script you used. 

![Image](https://github.com/user-attachments/assets/7b357652-6961-4db3-988e-9b23fd253bd2)

It would be also better if we could automatically merge the four operations in the new graph—reshape, unsqueeze, transpose, and squeeze—into just two, as I suggested here automatically from PyTorch.
https://github.com/m23ayou2/Pytorch_issue_mha_onnx/blob/main/optimize.ipynb

Other than that, I think the new computational graph is far better than before. 

Thanks a lot.

**Comment by justinchuby** [comment]
Are you using the latest PyTorch and onnxscript?

**Comment by m23ayou2** [comment]
using pytorch 2.7.1 and onnxscript 0.3.0, for this model archi: 

`
class mha_(nn.Module):
    def __init__(self):
        super(mha_, self).__init__()

        self.li = nn.Linear(embed_dim, embed_dim)

        self.ma = nn.MultiheadAttention(embed_dim=embed_dim, num_heads=num_heads, dropout=0, bias=False, add_bias_kv=False, add_zero_attn=False)

    def forward(self, x):
        li = self.li(x)
        q = li
        k = li
        v = li

        output, _ = self.ma(q, k, v)
        return output
`

![Image](https://github.com/user-attachments/assets/f3365c7c-552f-426a-879d-93f98baee47b)

**Comment by titaiwangms** [comment]
> The computational graph you have joined is substantially better than the first results and authentically reproduce what i was expecting, but could not reproduce it just using `torch.onnx.export(..., dynamo=True) ` If you can attach the script you used.

```python
import torch
import torch.nn as nn
import onnxruntime as ort
import math
import copy
seed = 25

torch.manual_seed(seed)
class MultiHeadAttention(nn.Module):
    def __init__(self, embed_dim, num_heads, mask=None):
        super(MultiHeadAttention, self).__init__()
        self.mask = mask
        self.embed_dim = embed_dim
        self.head_dim = embed_dim // num_heads
        self.num_heads = num_heads
        
        # Initialize weight matrices
        self.in_weights_q = nn.Parameter(torch.Tensor(embed_dim, embed_dim))
        self.in_weights_k = nn.Parameter(torch.Tensor(embed_dim, embed_dim))
        self.in_weights_v = nn.Parameter(torch.Tensor(embed_dim, embed_dim))
        
        self.out_weights = nn.Parameter(torch.Tensor(embed_dim, embed_dim))
        
        
    def forward(self, q, k , v):
        batch_size, seq_len, _ = q.shape
        
        # Project inputs to queries, keys and values
        q = torch.matmul(q, self.in_weights_q.T)   
        k = torch.matmul(k, self.in_weights_k.T)
        v = torch.matmul(v, self.in_weights_v.T)

        # Reshape for multi-head attention
        q = q.reshape(batch_size, seq_len * self.num_heads, self.head_dim).transpose(0, 1)
        k = k.reshape(batch_size, seq_len * self.num_heads, self.head_dim).transpose(0, 1)
        v = v.reshape(batch_size, seq_len * self.num_heads, self.head_dim).transpose(0, 1)
        
        # Calculate scaled dot-product attention
        scale = 1.0 / math.sqrt(self.head_dim)
        #print(q.shape, k.transpose(1, 2).shape)
        scores = torch.bmm(q, k.transpose(1, 2))
        attn = torch.softmax(scores * scale, dim=-1)
        #print(scores.shape)
        
        # Apply attention to values
        output = torch.bmm(attn, v)
        print(output.shape)
        
        # Reshape and project back to output dimension
        output = output.transpose(0, 1).reshape(seq_len * batch_size, self.embed_dim)
        output = output @ self.out_weights.T
        output = output.reshape(batch_size, seq_len, self.embed_dim)
        
        return output
# Example usage
embed_dim = 20
num_heads = 4
seq_len = 3
batch_size = 1
head_dim = int(embed_dim // num_heads)


input_tensor = torch.randn(batch_size, seq_len,  embed_dim)

multihead_attn = MultiHeadAttention(embed_dim=embed_dim, num_heads=num_heads)
mha = nn.MultiheadAttention(embed_dim=embed_dim, num_heads=num_heads, dropout=0, bias=False, add_bias_kv=False, add_zero_attn=False)

wq, wk, wv = torch.split(mha.in_proj_weight, [embed_dim, embed_dim, embed_dim], dim=0)
multihead_attn.in_weights_q.data = wq
multihead_attn.in_weights_k.data = wk
multihead_attn.in_weights_v.data = wv
multihead_attn.out_weights.data = mha.out_proj.weight
torch.abs(multihead_attn(input_tensor, input_tensor, input_tensor) / mha(input_tensor, input_tensor, input_tensor)[0]-1).max() * 100

class mha_(nn.Module): 
    def __init__(self): 
        super(mha_, self).__init__()

        self.li = nn.Linear(embed_dim, embed_dim)

        self.ma = MultiHeadAttention(embed_dim=embed_dim, num_heads=num_heads)

    def forward(self, x):
        li = self.li(x)
        q = li
        k = li
        v = li

        output = self.ma(q,k,v)
        return output
mha_c = mha_()
# Export the model to ONNX
torch.onnx.export(
    mha_c,                    
    input_tensor,             
    "mha_model.onnx",          
    input_names=['input'],   
    output_names=['output'],
    dynamo=True
)

print("Model successfully exported to ONNX format.")
torch.Size([12, 1, 5])

class mha_(nn.Module): 
    def __init__(self): 
        super(mha_, self).__init__()

        self.li = nn.Linear(embed_dim, embed_dim)

        self.ma = nn.MultiheadAttention(embed_dim=embed_dim, num_heads=num_heads, dropout=0, bias=False, add_bias_kv=False, add_zero_attn=False)

    def forward(self, x):
        li = self.li(x)
        q = li
        k = li
        v = li

        output, _ = self.ma(q, k, v)
        return output
mha_c = mha_()
# Export the model to ONNX
torch.onnx.export(
    mha_c,                    
    input_tensor,             
    "std_mha_model.onnx",          
    input_names=['input'],   
    output_names=['output'],
    dynamo=True)

print("Model successfully exported to ONNX format.")
```

> ![Image](https://github.com/user-attachments/assets/7b357652-6961-4db3-988e-9b23fd253bd2)
> 
> It would be also better if we could automatically merge the four operations in the new graph—reshape, unsqueeze, transpose, and squeeze—into just two, as I suggested here automatically from PyTorch. https://github.com/m23ayou2/Pytorch_issue_mha_onnx/blob/main/optimize.ipynb

You can do whatever post-processing you want with graph-level pass or even just simple surgeries through ONNX-IR (try to build/modify your own model.)
IR: https://github.com/onnx/ir-py/tree/main
Passes: https://github.com/onnx/ir-py/tree/main/src/onnx_ir/passes/common

> Other than that, I think the new computational graph is far better than before.
> 
> Thanks a lot.

**Comment by m23ayou2** [comment]
Thanks a lot, these are actually the scripts I made at first, I thought you made it possible directly through PyTorch.

**Comment by titaiwangms** [comment]
> Thanks a lot, these are actually the scripts I made at first, I thought you made it possible directly through PyTorch.

It's dynamo=True. It's purely PyTorch. 

The suggestion of using ONNX-IR is for exported onnx model, which you said you would like to "merge the four operations in the new graph—reshape, unsqueeze, transpose, and squeeze—into just two" -> Please check ONNX-IR to see how to play around with the ONNX model.
